### PR TITLE
[psbt] Convert non-witness UTXOs to witness if witness sig created

### PR DESCRIPTION
**Pull request by achow101**
If a witness signature was created when a non-witness UTXO is used, convert the non-witness UTXO to a witness one.

Port of #14196 to master.

### CODE REVIEW

File: src/script/sign.cpp [hunk 1]
```diff
@@ -276,6 +276,13 @@ bool SignPSBTInput(const SigningProvider& provider, const CMutableTransaction& t
     // Verify that a witness signature was produced in case one was required.
     if (require_witness_sig && !sigdata.witness) return false;
     input.FromSignatureData(sigdata);
+
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/script/sign.cpp]: Fixed

**Comment by gmaxwell** [comment]
ACK

**Comment by sipa** [comment]
utACK 52d9adcdfbf545cb9a9facb482cb3673578abfa0. I'll test this soon.

File: src/script/sign.cpp [hunk 2]
```diff
@@ -277,6 +277,13 @@ bool SignPSBTInput(const SigningProvider& provider, const CMutableTransaction& t
     if (require_witness_sig && !sigdata.witness) return false;
     input.FromSignatureData(sigdata);
 
+    if (sigdata.witness) {
+        // Convert the non-witness utxo to witness
+        if (input.witness_utxo.IsNull() && input.non_witness_utxo) {
+            input.witness_utxo = input.non_witness_utxo->vout[tx.vin[index].prevout.n];
```
Comment (promag) [inline src/script/sign.cpp]: IIUC this could be
```cpp
assert(!utxo.IsNull());
input.witness_utxo = utxo;
```

**Comment by sipa** [comment]
Here is a test that fails before and succeeds after this PR: https://github.com/sipa/bitcoin/commit/ce238f2aff3253cdde88d0f9b995b7e0016cb52b

**Comment by sipa** [comment]
@achow101 Feel free to cherry-pick the test into this PR.

That RPC test is a reproduction of what I did to notice the issue in the first place, so Tested ACK from me.

File: test/functional/rpc_psbt.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 """
 
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_equal, assert_raises_rpc_error, find_output
+from test_framework.util import assert_equal, assert_raises_rpc_error, find_output, disconnect_nodes, connect_nodes_bi, sync_blocks
```
Comment (promag) [inline test/functional/rpc_psbt.py]: nit, nit, nit, sort.

**Review by promag — COMMENTED** [review]
utACK 862d159.

**Comment by laanwj** [comment]
utACK 862d159d635c1de219d94e030b186a745fe28eb9

**Comment by fanquake** [comment]
862d159 will be backported in #14780.

**Pull request merged**